### PR TITLE
DATAGO-134167: Native Insights Agent Pro (third-party forwarding) support

### PR DESCRIPTION
**Pull request by hamidp555**
## Summary

Adds native chart support for the Solace Insights **forwarding (Agent Pro)** modes, so a single `helm install -f values.yaml` covers standard, forwarding-only, and forwarding + Datadog-push deployments with no out-of-band `kubectl create secret` / `kubectl patch` steps, and `helm upgrade` keeps the config in place.

Jira: DATAGO-134167 (also closes DATAGO-134542 — apply `insights.image.pullPolicy`).

## Changes

- **`insights.forwarding.{enabled, otelConfig, logsConfig}`**
  - `otelConfig` → chart-managed Secret mounted at `/etc/datadog-agent/otel-config.yaml` (required when `forwarding.enabled=true`).
  - `logsConfig` → chart-managed ConfigMap mounted over `/etc/datadog-agent/conf.d/solace.d/logs.yml` (only when `forwarding.enabled=true`).
  - Sets `SOLACE_CUSTOM_INSIGHTS_ENABLED=true` and `INSIGHTS_AGENT_TELEMETRY_ENABLED=false` in forwarding mode.
  - Datadog dual-write (`INSIGHTS_AGENT_ADDITIONAL_ENDPOINTS` / `..._LOGS_CONFIG_ADDITIONAL_ENDPOINTS`) and `INSIGHTS_AGENT_GOMEMLIMIT` are operator-supplied via `insights.environmentVariables` (passthrough).
- **insights-agent resource limits**: used verbatim when set; otherwise computed as the requests value plus a per-`solace.size` OTel headroom (`_helpers.tpl`). `requests` default to `256Mi`/`200m`. Mi/Gi and cores/millicores parsed via `float64`; unsupported memory units fail fast.
- **`insights.image.pullPolicy`** (default `Always`) now applied to the agent container (DATAGO-134542).
- **Fail-fast validation**: `forwarding.enabled` requires `otelConfig`; `logsConfig` only takes effect with `forwarding.enabled`.
- Updated `values.schema.json`, `NOTES.txt`, `INSIGHTS.md`; bumped chart to **3.10.0**.

## Testing

- `helm lint` clean.
- New `tests/python/unit/test_insights_forwarding.py` + updated `test_insights.py`; full unit suite passes (37 tests), covering all three modes, gating, validation, resource-limit computation, and unit-parsing edge cases.

## Notes

- When `solace.systemScaling` is used (so `solace.size` is ignored for broker sizing), the computed agent headroom falls back to the `solace.size` default tier — set `insights.resources.limits` explicitly to size the agent for a custom-scaled broker (documented in `values.yaml`/`INSIGHTS.md`).

### CODE REVIEW

File: pubsubplus/Chart.yaml [hunk 1]
```diff
@@ -1,7 +1,7 @@
 apiVersion: v2
 description: Deploy Solace Event Broker Singleton or HA redundancy group onto a Kubernetes Cluster
 name: pubsubplus
-version: 3.9.0
```
Comment (Ali-Zmn) [inline pubsubplus/Chart.yaml:4]: should we go 3.10 or 4 with this change?

**Review by Ali-Zmn — APPROVED** [review]

File: pubsubplus/Chart.yaml (hunk 1 above)
Comment (hamidp555) [reply pubsubplus/Chart.yaml:4]: I think 3.10.0 should be fined since insights is a small part of this helm chart  + no breaking changes are introduced

File: pubsubplus/templates/insights-agent-secret.yaml [hunk 2]
```diff
@@ -69,4 +94,8 @@ stringData:
   INSIGHTS_AGENT_SEMP_PORT: "{{ template "insightsAgent.sempPort" . }}"
   INSIGHTS_AGENT_SEMP_PROTOCOL: {{ template "insightsAgent.sempProtocol" . }}
   INSIGHTS_AGENT_HEALTH_CHECK_PORT: "5550"
+{{- if .Values.insights.forwarding.enabled }}
+  SOLACE_CUSTOM_INSIGHTS_ENABLED: "true"
+  INSIGHTS_AGENT_TELEMETRY_ENABLED: "false"
+{{- end }}
```
[the remark this answers is not in the document]
Comment (hamidp555) [reply pubsubplus/templates/insights-agent-secret.yaml:125]: Addressed in 399fe6e. The `data` loop now excludes all chart-managed `stringData` keys — the SEMP/hostname/health keys always, plus `SOLACE_CUSTOM_INSIGHTS_ENABLED` and `INSIGHTS_AGENT_TELEMETRY_ENABLED` in forwarding mode — so each appears only once (in `stringData`). Operator-supplied custom vars still pass through to `data`. Added `test_chart_managed_keys_not_duplicated_in_data`.

File: tests/python/unit/test_insights_forwarding.py [hunk 3]
```diff
@@ -0,0 +1,414 @@
...
+    resources = render_helm_template(values)
+
+    # Chart-managed OTel config Secret is rendered (keyed per pod index).
+    otel_secret = _otel_secret(resources)
+    assert otel_secret is not None
+    assert "otel-config-0.yaml" in otel_secret["stringData"]
+
+    # Only SOLACE_CUSTOM_INSIGHTS_ENABLED is chart-managed in stringData.
+    env_secret = _env_secret(resources)
+    assert env_secret["stringData"]["SOLACE_CUSTOM_INSIGHTS_ENABLED"] == "true"
```
[the remark this answers is not in the document]
Comment (hamidp555) [reply tests/python/unit/test_insights_forwarding.py]: Addressed in 0c964f1. The test now asserts `INSIGHTS_AGENT_TELEMETRY_ENABLED == \"false\"` alongside `SOLACE_CUSTOM_INSIGHTS_ENABLED`, and the stale comment was corrected.

File: pubsubplus/templates/solaceStatefulSet.yaml [hunk 4]
```diff
@@ -295,6 +296,19 @@ spec:
           - name: data
             mountPath: /opt/datadog-agent/run
             subPath: opt/datadog-agent/run
+{{- if .Values.insights.forwarding.enabled }}
+          - name: otel-config
+            mountPath: /etc/datadog-agent/otel-config.yaml
+            # Each node mounts its own config by pod index (primary=0, backup=1, monitor=2).
+            subPathExpr: otel-config-$(INSIGHTS_AGENT_NODE_ROLE).yaml
+            readOnly: true
```
[the remark this answers is not in the document]
Comment (hamidp555) [reply pubsubplus/templates/solaceStatefulSet.yaml:313]: Addressed in ed20545. Switched to a version-independent source: added a `POD_NAME` env (`fieldRef: metadata.name`, present on all k8s versions), the Secret keys are now `otel-config-<fullname>-<ordinal>.yaml`, and the mount uses `subPathExpr: otel-config-$(POD_NAME).yaml`. No longer depends on the `apps.kubernetes.io/pod-index` label.

**Review by cybtron — APPROVED** [review]
Approving — validated the latest revision addresses all earlier review concerns:

- **Unknown `solace.size`**: the limit helpers now `fail` with an actionable message instead of silently defaulting the OTel headroom to zero (`hasKey` guard in `_helpers.tpl`). Covered by `test_agent_limits_unknown_size_fails_when_computed` and `_ok_with_explicit_limits`.
- **Per-node OTel config keying**: re-keyed from the `apps.kubernetes.io/pod-index` label to `POD_NAME` (`metadata.name`), which is present on all K8s versions and removes the fragile numeric-vs-string coupling with the entrypoint. Secret key (`otel-config-<fullname>-<idx>.yaml`) and `subPathExpr` line up.
- **Telemetry contract**: forwarding mode now asserts `INSIGHTS_AGENT_TELEMETRY_ENABLED=false`.

Tests cover all three modes, fail-fast validation, resource-limit computation, and unit-parsing edge cases. Backward compatibility (standard mode no-op) is preserved.

Note: this needs to ship together with the agent-side entrypoint change (solace-datadog-agent#472) — confirm the default agent image tag includes it.

LGTM.

File: pubsubplus/templates/_helpers.tpl [hunk 5]
```diff
@@ -38,3 +38,42 @@ Determine the service type based on redundancy
...
+{{- if not (hasKey $deltas .Values.solace.size) -}}
+{{- fail (printf "insights: cannot compute the agent cpu limit for unknown solace.size %q; set insights.resources.limits.cpu explicitly" .Values.solace.size) -}}
+{{- end -}}
+{{- $delta := int (get $deltas .Values.solace.size) -}}
+{{- $requests := .Values.insights.resources.requests | default dict -}}
+{{- $base := $requests.cpu | default "200m" | toString -}}
+{{- $baseM := 0.0 -}}
+{{- if hasSuffix "m" $base -}}{{- $baseM = trimSuffix "m" $base | float64 -}}
+{{- else -}}{{- $baseM = mulf ($base | float64) 1000.0 -}}{{- end -}}
+{{- divf (add (ceil $baseM | int) $delta) 1000 -}}
```
Comment (LewisKSaint) [inline pubsubplus/templates/_helpers.tpl]: nit: this may generate a floats

**Review by LewisKSaint — APPROVED** [review]

**Comment by hamidp555** [comment]
https://sol-jira.atlassian.net/wiki/spaces/MI/pages/6406504488/Self-Managed+QuickStart+Helm+with+Third-Party-Forwarding page shows the instruction on how to deploy insights agent with forwarding enabled

File: pubsubplus/templates/_helpers.tpl (hunk 5 above)
Comment (hamidp555) [reply pubsubplus/templates/_helpers.tpl]: Intentional — Kubernetes accepts fractional CPU cores (`0.7`, `2.2`, equivalent to `700m`/`2200m`), and we render the cores form by design. `$baseM` is a float only momentarily: it is re-`int`-ed via `ceil … | int` before the delta is added, so the only float is the final `divf` (the cores value we want to emit). Go renders float64 in shortest round-tripping form, so it prints `0.7`, not `0.69999…`. Verified renders: dev/prod1k `0.7`, prod10k `1.2`, prod100k/prod200k `2.2` — all valid CPU quantities.
